### PR TITLE
Initial CodeLens Provider

### PR DESCRIPTION
**Pull request by awschristou**
Seeding the CodeLens feature branch

This is the initial support for #135 and #136

### CODE REVIEW

File: src/shared/templates/sam/samTemplateGenerator.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        }
+
+        return {
+            Resources: resources
+        }
+    }
+
+    private async createTemplateFromExistingTemplate(): Promise<CloudFormationTemplate> {
+        const template: CloudFormationTemplate = await this.loadTemplate(this._existingTemplateFilename!)
+        const resource: CloudFormationResource = template.Resources![this._resourceName!]
```
[the remark this answers is not in the document]
Comment (awschristou) [reply src/shared/templates/sam/samTemplateGenerator.ts]: `validate` is intended to make all of the assurances up front

File: tsconfig.json [hunk 2]
```diff
@@ -1,12 +1,17 @@
...
-        "paths": { "*": ["types/*"] },
+        "paths": {
+            "*": [
+                "types/*"
+            ]
+        },
         "module": "commonjs",
         "target": "es6",
         "outDir": "out",
         "lib": [
```
Comment (mpiroc) [inline tsconfig.json]: We can remove the `lib` section. If `lib` does not exist, and `target` is `es6`, the following libs will be injected:

* `DOM`
* `ES6`
* `DOM.Iterable`
* `ScriptHost`

See https://www.typescriptlang.org/docs/handbook/compiler-options.html

File: src/shared/sam/packagers/samPackager.ts [hunk 3]
```diff
@@ -81,7 +81,7 @@ export class SamPackager {
     private async validateFoldersExist(): Promise<void> {
         const missingFolders: boolean[] = (await Promise.all(
             Object.keys(this._folderMap)
-                .map(folder => SystemUtilities.fileExists(folder))
```
Comment (mpiroc) [inline src/shared/sam/packagers/samPackager.ts]: Do you also need to `await` the result of `fileExists`?

Comment (awschristou) [reply src/shared/sam/packagers/samPackager.ts]: No, I'm returning them all, then calling Promise.all

**Pull request merged**